### ISSUE TITLE
Problem with Nunchaku inference of SVDQuant produced SANA model

### ISSUE THREAD

**Issue by bogdanovicandjela** [nunchaku-tech/nunchaku#488]
**Motivation**

I tried to create a function to convert a quantized SANA model from DeepCompressor to Nunchaku (not supported yet), based on the conversion of the FLUX model and the quantized int4 SANA model from Hugging Face. [https://huggingface.co/mit-han-lab/svdq-int4-sana-1600m/tree/main](url)

Model was quantized using SVDQuant method, after quantization the model was validated by Deepcompressor eval routine and produced meaningful output. After int4 quantization in DeepCompressor, a transformer block from model.pt looks like this:

![Image](https://github.com/user-attachments/assets/0d5e8969-a20b-4500-ac31-0f0eda9b1b8f)

After conversion to Nunchaku, a transformer block looks like this:

![Image](https://github.com/user-attachments/assets/7d58d93a-1f3c-4f21-8c60-d28a32d8aed0)

![Image](https://github.com/user-attachments/assets/af9244a7-9684-4bd6-8e8f-9ba75dd546a1)

In Nunchaku I tried to load model(.safetensor format) and the difference between our model and HF model ( [https://huggingface.co/mit-han-lab/svdq-int4-sana-1600m/tree/main](url)) is in the dimensions of the "qkv_proj" layer: as shown in the previous image, the resulting tensor has dimensions [6784, 1152], but in the HF model dimensions are [6912, 1152].

The error message I get in Nunchaku when running inference is the following:

![Image](https://github.com/user-attachments/assets/b8b50305-20a5-4978-a5ea-12dae2847d5b)

I also tried using padding to match the expected output dimensions, but in that case, for the same prompt, the HF model produces the image on the right, while our model generates the image on the left. (All tensors have the same dimensions, names and types, and the padded zeros are in the same positions in both our model and the HF model)

![Image](https://github.com/user-attachments/assets/2595af3c-a9f8-4217-b5ff-8a91614a9dff)

**Related resources**

New functions in convert.py script from deepcompressor/backend/nunchaku are:

- **convert_to_nunchaku_sana_state_dicts**
- **convert_to_nunchaku_sana_transformer_block_state_dict**

And I also made changes in **convert_to_nunchaku_w4x4y16_linear_state_dict** and **convert_to_nunchaku_transformer_block_state_dict** at specified part.
```
def convert_to_nunchaku_sana_state_dicts(

    state_dict: dict[str, torch.Tensor],
    scale_dict: dict[str, torch.Tensor],
    smooth_dict: dict[str, torch.Tensor],
    branch_dict: dict[str, torch.Tensor],
    float_point: bool = False,
) -> tuple[dict[str, torch.Tensor], dict[str, torch.Tensor]]:
    block_names: set[str] = set()
    other: dict[str, torch.Tensor] = {}
    for param_name in state_dict.keys():
        if param_name.startswith(("transformer_blocks.")):
            block_names.add(".".join(param_name.split(".")[:2]))  # First two parts are block name, maybe more?
        else:
            other[param_name] = state_dict[param_name]
    block_names = sorted(block_names, key=lambda x: (x.split(".")[0], int(x.split(".")[-1])))
    print(f"Converting {len(block_names)} transformer blocks...")
    converted: dict[str, torch.Tensor] = {}
    for block_name in block_names:
        if block_name.startswith("transformer_blocks"):
            convert_fn = convert_to_nunchaku_sana_transformer_block_state_dict
        update_state_dict(
            converted,
            convert_fn(
                state_dict=state_dict,
                scale_dict=scale_dict,
                smooth_dict=smooth_dict,
                branch_dict=branch_dict,
                block_name=block_name,
                float_point=float_point,
            ),
            prefix=block_name,
        )
    return converted, other
```

```
def convert_to_nunchaku_sana_transformer_block_state_dict(
    state_dict: dict[str, torch.Tensor],
    scale_dict: dict[str, torch.Tensor],
    smooth_dict: dict[str, torch.Tensor],
    branch_dict: dict[str, torch.Tensor],
    block_name: str,
    float_point: bool = False,
) -> dict[str, torch.Tensor]:
    conv_point_local_name = "ff.conv_point"
    if f"{block_name}.{conv_point_local_name}.weight" not in state_dict:
        conv_point_local_name = "ff"
        assert f"{block_name}.{conv_point_local_name}.weight" in state_dict
    conv_inverted_local_name = "ff.conv_inverted"
    if f"{block_name}.{conv_inverted_local_name}.weight" not in state_dict:
        conv_inverted_local_name = "ff"
        assert f"{block_name}.{conv_inverted_local_name}.weight" in state_dict

    return convert_to_nunchaku_transformer_block_state_dict(
        state_dict=state_dict,
        scale_dict=scale_dict,
        smooth_dict=smooth_dict,
        branch_dict=branch_dict,
        block_name=block_name,
        local_name_map={

            "attn.qkv_proj": ["attn1.to_q", "attn1.to_k", "attn1.to_v"],
            "cross_attn.kv_linear": ["attn2.to_k", "attn2.to_v"], 
            "cross_attn.q_linear": ["attn2.to_q"],
            "attn.out_proj": "attn1.to_out.0", 
            "cross_attn.out_proj": "attn2.to_out.0", 
            "ff.depth_conv": "ff.conv_depth", 
            "ff.inverted_conv": conv_inverted_local_name,
            "ff.point_conv": conv_point_local_name,
            "scale_shift_table": "scale_shift_table",
        },
        smooth_name_map={
            "attn.qkv_proj": "attn1.to_q",
            "cross_attn.q_linear": "attn2.to_q",
            "attn.out_proj": "attn1.to_out.0",
            "cross_attn.out_proj": "attn2.to_out.0", 
            "ff.inverted_conv": conv_inverted_local_name,
            "ff.point_conv": conv_point_local_name,
        },
        branch_name_map={
            "attn.qkv_proj": "attn1.to_q",
            "cross_attn.q_linear": "attn2.to_q",
            "attn.out_proj": "attn1.to_out.0",
            "cross_attn.out_proj": "attn2.to_out.0", 
            "ff.inverted_conv": conv_inverted_local_name,
            "ff.point_conv": conv_point_local_name,
        },
        convert_map={
            "attn.qkv_proj": "linear",
            "cross_attn.kv_linear": "kv_proj", 
            "cross_attn.q_linear": "linear",
            "attn.out_proj": "linear",
            "cross_attn.out_proj": "linear",
            "ff.depth_conv": "dwconv",
            "ff.inverted_conv": "linear",
            "ff.point_conv": "linear",
            "scale_shift_table": "scale_table",
        },
        float_point=float_point,
    )

```

```
def convert_to_nunchaku_w4x4y16_linear_state_dict(
    weight: torch.Tensor,
    scale: torch.Tensor,
    bias: torch.Tensor | None = None,
    smooth: torch.Tensor | None = None,
    lora: tuple[torch.Tensor, torch.Tensor] | None = None,
    shift: torch.Tensor | None = None,
    smooth_fused: bool = False,
    float_point: bool = False,
    subscale: torch.Tensor | None = None,
) -> dict[str, torch.Tensor]:
    if weight.ndim > 2:  # pointwise conv
        assert weight.numel() == weight.shape[0] * weight.shape[1]
        weight = weight.view(weight.shape[0], weight.shape[1])
    if scale.numel() > 1:
       

>  if scale.ndim > 4: # pointwise conv
>             assert scale.numel() == scale.shape[0] * scale.shape[2]
>             scale = scale.view(scale.shape[0], 1, scale.shape[2], 1)

        assert scale.ndim == weight.ndim * 2
        assert scale.numel() == scale.shape[0] * scale.shape[2]
        scale = scale.view(scale.shape[0], 1, scale.shape[2], 1)
        scale_key = "wcscales" if scale.shape[2] == 1 else "wscales"
    else:
        scale_key = "wtscale"
    if subscale is None:
        subscale_key = ""
    else:
        assert subscale.ndim == weight.ndim * 2
        assert subscale.numel() == subscale.shape[0] * subscale.shape[2]
        assert subscale.numel() > 1
        subscale = subscale.view(subscale.shape[0], 1, subscale.shape[2], 1)
        subscale_key = "wcscales" if subscale.shape[2] == 1 else "wscales"
    if lora is not None and (smooth is not None or shift is not None):
        # unsmooth lora down projection
        dtype = weight.dtype
        lora_down, lora_up = lora
        lora_down = lora_down.to(dtype=torch.float64)
        if smooth is not None and not smooth_fused:
            lora_down = lora_down.div_(smooth.to(torch.float64).unsqueeze(0))
        if shift is not None:
            bias = torch.zeros([lora_up.shape[0]], dtype=torch.float64) if bias is None else bias.to(torch.float64)
            if shift.numel() == 1:
                shift = shift.view(1, 1).expand(lora_down.shape[1], 1).to(torch.float64)
            else:
                shift = shift.view(-1, 1).to(torch.float64)
            bias = bias.add_((lora_up.to(dtype=torch.float64) @ lora_down @ shift).view(-1))
            bias = bias.to(dtype=dtype)
        lora = (lora_down.to(dtype=dtype), lora_up)
    weight, scale, bias, smooth, lora, subscale = convert_to_nunchaku_w4x4y16_linear_weight(
        weight, scale=scale, bias=bias, smooth=smooth, lora=lora, float_point=float_point, subscale=subscale
    )
    state_dict: dict[str, torch.Tensor] = {}
    state_dict["qweight"] = weight
    state_dict[scale_key] = scale
    if subscale is not None:
        state_dict[subscale_key] = subscale
    state_dict["bias"] = bias
    #state_dict["smooth_orig"] = smooth
    state_dict["smooth"] = torch.ones_like(smooth) if smooth_fused else smooth.clone()
    if lora is not None:
        state_dict["lora_down"] = lora[0]
        state_dict["lora_up"] = lora[1]
    return state_dict
```

```
def convert_to_nunchaku_transformer_block_state_dict(
    state_dict: dict[str, torch.Tensor],
    scale_dict: dict[str, torch.Tensor],
    smooth_dict: dict[str, torch.Tensor],
    branch_dict: dict[str, torch.Tensor],
    block_name: str,
    local_name_map: dict[str, str | list[str]],
    smooth_name_map: dict[str, str],
    branch_name_map: dict[str, str],
    convert_map: dict[str, str],
    float_point: bool = False,
) -> dict[str, torch.Tensor]:
    print(f"Converting block {block_name}...")
    converted: dict[str, torch.Tensor] = {}
    candidates: dict[str, torch.Tensor] = {
        param_name: param for param_name, param in state_dict.items() if param_name.startswith(block_name)
    }
    for converted_local_name, candidate_local_names in tqdm.tqdm(
        local_name_map.items(), desc=f"Converting {block_name}", dynamic_ncols=True
    ):
        if isinstance(candidate_local_names, str):
            candidate_local_names = [candidate_local_names]

>         if "scale_shift_table" in candidate_local_names:
>             if convert_map[converted_local_name] == "scale_table":
>                 print(f"  - Converting {block_name} scale_shift_table of {candidate_local_names} to {converted_local_name}.")
>                 converted[f"{converted_local_name}"] = candidates[f"{block_name}.{candidate_local_names[0]}"].clone().cpu()
>         else:
>             candidate_names = [f"{block_name}.{candidate_local_name}" for candidate_local_name in candidate_local_names]
>             weight = [candidates[f"{candidate_name}.weight"] for candidate_name in candidate_names]
>             bias = [candidates.get(f"{candidate_name}.bias", None) for candidate_name in candidate_names]
>             scale = [scale_dict.get(f"{candidate_name}.weight.scale.0", None) for candidate_name in candidate_names]
>             subscale = [scale_dict.get(f"{candidate_name}.weight.scale.1", None) for candidate_name in candidate_names]
>             print(f"CANDIDATES: {candidate_names}")
>             if len(weight) > 1:
>                 bias = None if all(b is None for b in bias) else torch.concat(bias, dim=0)
>                 if all(s is None for s in scale):
>                     scale = None
>                 else:
>                     if scale[0].numel() == 1:  # switch from per-tensor to per-channel scale
>                         assert all(s.numel() == 1 for s in scale)
>                         scale = torch.concat(
>                             [
>                                 s.view(-1).expand(weight[i].shape[0]).reshape(weight[i].shape[0], 1, 1, 1)
>                                 for i, s in enumerate(scale)
>                             ],
>                             dim=0,
>                         )
>                     else:
>                         scale = torch.concat(scale, dim=0)
>                 subscale = None if all(s is None for s in subscale) else torch.concat(subscale, dim=0)
>                 weight = torch.concat(weight, dim=0)
>             else:
>                 weight, bias, scale, subscale = weight[0], bias[0], scale[0], subscale[0]
>             smooth = smooth_dict.get(f"{block_name}.{smooth_name_map.get(converted_local_name, '')}", None)
>             branch = branch_dict.get(f"{block_name}.{branch_name_map.get(converted_local_name, '')}", None)
>             if branch is not None:
>                 branch = (branch["a.weight"], branch["b.weight"])
>             if scale is None and converted_local_name != 'ff.depth_conv':
>                 #assert smooth is None and branch is None and subscale is None
>                 print(f"  - Copying {block_name} weights of {candidate_local_names} as {converted_local_name}.weight")
>                 converted[f"{converted_local_name}.weight"] = weight.clone().cpu()
>                 if bias is not None:
>                     print(f"  - Copying {block_name} biases of {candidate_local_names} as {converted_local_name}.bias")
>                     converted[f"{converted_local_name}.bias"] = bias.clone().cpu()
>                 continue
>             if convert_map[converted_local_name] == "adanorm_single":
>                 print(f"  - Converting {block_name} weights of {candidate_local_names} to {converted_local_name}.")
>                 update_state_dict(
>                     converted,
>                     convert_to_nunchaku_w4x16_adanorm_single_state_dict(weight=weight, scale=scale, bias=bias),
>                     prefix=converted_local_name,
>                 )
>             elif convert_map[converted_local_name] == "adanorm_zero":
>                 print(f"  - Converting {block_name} weights of {candidate_local_names} to {converted_local_name}.")
>                 update_state_dict(
>                     converted,
>                     convert_to_nunchaku_w4x16_adanorm_zero_state_dict(weight=weight, scale=scale, bias=bias),
>                     prefix=converted_local_name,
>                 )
>             elif convert_map[converted_local_name] == "dwconv":
>                 print(f"  - Converting {block_name} DWCONV of {candidate_local_names} to {converted_local_name}.")
>                 weight = weight.permute(0, 2, 3, 1)
>                 converted[f"{converted_local_name}.weight"] = weight.clone().cpu()
>                 if bias is not None:
>                     print(f"  - Copying {block_name} biases of {candidate_local_names} as {converted_local_name}.bias")
>                     converted[f"{converted_local_name}.bias"] = bias.clone().cpu()
>             elif convert_map[converted_local_name] == "kv_proj":
>                 print(f"  - Converting {block_name} kv_proj of {candidate_local_names} to {converted_local_name}.weight")
>                 converted[f"{converted_local_name}.weight"] = weight.clone().cpu()    
>                 if bias is not None:
>                     print(f"  - Copying {block_name} biases of {candidate_local_names} as {converted_local_name}.bias")
>                     converted[f"{converted_local_name}.bias"] = bias.clone().cpu()
>             elif convert_map[converted_local_name] == "linear":
>                 smooth_fused = "out_proj" in converted_local_name and smooth_dict.get("proj.fuse_when_possible", True)
>                 shift = [candidates.get(f"{candidate_name[:-7]}.shift", None) for candidate_name in candidate_names]
>                 assert all(s == shift[0] for s in shift)
>                 shift = shift[0]
>                 print(
>                     f"  - Converting {block_name} weights of {candidate_local_names} to {converted_local_name}."
>                     f" (smooth_fused={smooth_fused}, shifted={shift is not None}, float_point={float_point})"
>                 )
>                 update_state_dict(
>                     converted,
>                     convert_to_nunchaku_w4x4y16_linear_state_dict(
>                         weight=weight,
>                         scale=scale,
>                         bias=bias,
>                         smooth=smooth,
>                         lora=branch,
>                         shift=shift,
>                         smooth_fused=smooth_fused,
>                         float_point=float_point,
>                         subscale=subscale,
>                     ),
>                     prefix=converted_local_name,
>                 )
>             else:
>                 raise NotImplementedError(f"Conversion of {convert_map[converted_local_name]} is not implemented.")
> 

    return converted
```

**Question**

Does anyone know what went wrong here?

**Comment by bubonja** [comment]
I was also looking for the SANA conversion script. Any information if the script will be released? Also pointers on how to solve this issue would be useful as well. 🙂

**Comment by bogdanovicandjela** [comment]
I came across this issue [Nunchaku Summer Development Roadmap](https://github.com/mit-han-lab/nunchaku/issues/431) and saw that @synxlin is working on this topic. Could you take a look at my issue?

**Comment by lantudou** [comment]
Just to offer a piece of advice based on my experience, I encountered this same error when I tried to use DeepComposer to optimize our own Flux Mini model.


<img width="2173" height="147" alt="Image" src="https://github.com/user-attachments/assets/9b92a4ef-1e8b-4d2e-a57e-8727022ba909" /> The root cause is that the number of layers and features are hardcoded in Nunchaku's C source code. I was able to solve the problem by making the following changes.


<img width="2584" height="1702" alt="Image" src="https://github.com/user-attachments/assets/97d7d5cc-ed22-461e-a5ed-cf33f9f566e7" />

**Comment by khosravipasha** [comment]
I had a similar issue with using deepcompressor and SANA 1.6B, did the solution work? do we need to pad the model before giving it to deep compressor?

In my first try I did deepcompressor and then did padding after to match what the inference engine expects but that produces images like above.
